src/esp/gfx_batch/Hbao.cpp

```diff
@@ -0,0 +1,1069 @@
+ */
+struct HbaoUniformData {
+  Mn::Float radiusToScreen;
+  Mn::Float r2;
+  Mn::Float negInvR2;
+  Mn::Float nDotVBias;
+
+  Mn::Vector2 invFullResolution;
+  Mn::Vector2 invQuarterResolution;
+
+  Mn::Float aoMultiplier;
+  Mn::Float powExponent;
+  // padding
+  Mn::Int : 32;
+  Mn::Int : 32;
+
+  Mn::Vector4 projInfo;
+
+  Mn::Vector2 projScale;
+  Mn::Int projOrtho;
+  Mn::Int _pad1;
+
+  Mn::Vector4 float2Offsets[AoRandomTextureSize * AoRandomTextureSize];
+  Mn::Vector4 jitters[AoRandomTextureSize * AoRandomTextureSize];
+};
+
+static_assert(sizeof(HbaoUniformData) % 4 == 0, "Not a nice uniform struct");
+
+struct Hbao::State {
+  Mn::GL::Texture2D sceneDepthLinear;
+  Mn::GL::Framebuffer depthLinear{Mn::NoCreate};
+
+  Mn::GL::Texture2D sceneViewNormal;
+  Mn::GL::Framebuffer viewNormal{Mn::NoCreate};
+
+  Mn::GL::Texture2D hbaoResult;
+  Mn::GL::Texture2D hbaoBlur;
+  Mn::GL::Texture2DArray hbaoRandom;
+  Mn::GL::Framebuffer hbaoCalc{Mn::NoCreate};
+  Mn::GL::Framebuffer hbao2Deinterleave{Mn::NoCreate};
+  Mn::GL::Framebuffer hbao2Calc{Mn::NoCreate};
+
+  Mn::GL::Texture2DArray hbao2DepthArray;
+  Mn::GL::Texture2DArray hbao2ResultArray;
+
+  HbaoBlurShader bilateralBlurShader{/*bilateral*/ true, 0};
+  HbaoBlurShader hbaoBlurShader{/*bilateral*/ false, /*aoBlur2ndPass*/ 0};
+  HbaoBlurShader hbaoBlur2Shader{/*bilateral*/ false, /*aoBlur2ndPass*/ 1};
+  DepthLinearizeShader depthLinearizeShader{false};
+  DepthLinearizeShader depthLinearizeShaderMsaa{true};
+  ViewNormalShader viewNormalShader;
+  HbaoCalcShader hbaoCalcShader{/*deinterleaved*/ false,
+                                /*blur*/ false,
+                                {},
+                                false};
+  HbaoCalcShader hbaoCalcBlurShader{/*deinterleaved*/ false,
+                                    /*blur*/ true,
+                                    {},
+                                    false};
+  /* These depend on config, are created in the constructor */
+  HbaoCalcShader hbao2CalcShader{Mn::NoCreate};
+  HbaoCalcShader hbao2CalcBlurShader{Mn::NoCreate};
+  HbaoDeinterleaveShader hbao2DeinterleaveShader;
+  HbaoReinterleaveShader hbao2ReinterleaveShader{/*blur*/ false};
+  HbaoReinterleaveShader hbao2ReinterleaveBlurShader{/*blur*/ true};
+
+  Mn::GL::Buffer hbaoUniform{Mn::GL::Buffer::TargetHint::Uniform};
+  HbaoUniformData hbaoUniformData;
+
+  Mn::GL::Mesh triangle, triangleLayered;
+
+  HbaoConfiguration configuration;
+
+  Mn::Vector4 random[HbaoRandomNumElements * MaxSamples];
+};
+
+Hbao::Hbao(Mn::NoCreateT) noexcept {}
+
+Hbao::Hbao(const HbaoConfiguration& configuration) {
+  if (!Cr::Utility::Resource::hasGroup("gfx-batch-shaders")) {
+    importShaderResources();
+  }
+
+  state_.emplace();
+  setConfiguration(configuration);
+}
+
+Hbao::Hbao(Hbao&&) noexcept = default;
+
+Hbao::~Hbao() = default;
+
+Hbao& Hbao::operator=(Hbao&&) noexcept = default;
+
+void Hbao::setConfiguration(const HbaoConfiguration& configuration) {
```

Comment:
Wait a moment, I thought we agreed that you'd just replace the instance when configuration changes? This setter is only adding extra potential for bugs as it may not reininitialize all internal state properly.

In other words, this is what I mean:

```cpp
Hbao hbao;

...

// when settings change
hbao = Hbao{newHbaoConfiguration};
```

That way the instance is, when constructed, always correct, and there's no chance some state may be stale because `setConfiguration()` forgot to overwrite something that was only populated in a constructor.

For changing actual runtime parameters (as we discussed), there should be direct setters instead, separate from the `HbaoConfiguration` class which should have only the initial setup.

Reply:
I'll rework this when I work on the user-driven attributes-based customization functionality.